text/000-buildpacks-contribute-default-process-type.md

```diff
@@ -0,0 +1,149 @@
+
+#### Overrides - edge cases
+[edge-cases]: #edge-cases
+
+Example: multi-process app with an earlier buildpack declaring a process type of `web` with `[[processes.default]]` and a later buildpack redefining what `web` means WITHOUT `[[processes.default]]` -> the app image will have no default process. Users must specify the process to run at runtime.
+
+Example: multi-process app with an earlier buildpack declaring a process type of `web` with `[[processes.default]]` and a later buildpack redefining what `web` means with `[[processes.default]]` -> the app image will have a default process type of `web`, with the latest definition
+
+Example: multi-process app with an earlier buildpack declaring a process type of `web` with `[[processes.default]]` and a later buildpack redefining what `web` means with `override = false` -> the app image will have a default process type of `web` that is the earlier buildpack's definition.
+
+- Define the target persona: buildpack author, buildpack user.
+
+New buildpack users would need to be taught that buildpacks will be responsible for setting a sensible default process type. They should know how to override the buildpacks-provided default if they desire.
+
+Existing buildpack users would need to be taught about the change so that (1) they no longer have to pass a flag to specify the default process type or (2) they now need to pass a flag to explicitly override the default that buildpacks choose. For (2), it's likely that the number of users in this category would be small given that the lifecycle (pre- platform API 0.4) and pack (post- platform API 0.4) both encourage `web` as the default, and it's likely that buildpacks would also encourage this same default.
+
+Buildpack authors would need to be taught how to configure the default process type.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+The changes from this RFC should require a bump to the buildpack API.
+
+Currently, during the `build` phase, buildpacks may declare process types by writing entries to `<layers>/launch.toml`. During the `export` phase, the `<layers>/launch.toml` from each buildpack are aggregated to produce a combined processes list "such that process types from later buildpacks override identical process types from earlier buildpacks" ([spec](https://github.com/buildpacks/spec/blob/main/buildpack.md#process-3)).
+
+This RFC proposes a new boolean `override` key and a new nested table `[[processes.default]]` in `<layers>/launch.toml`:
+
+```toml
+[[processes]]
+type = "<process type>"
+command = "<command to run>"
+override = <boolean (defaults to true)>
+  [[processes.default]]
+  override = <boolean (defaults to true)>
+```
+
+Example:
+
+```toml
+[[processes]]
+type = web
+command = bundle exec ruby app.rb
+  [[processes.default]]
+  override = true
+
+[[processes]]
+type = worker
+command = bundle exec ruby worker.rb
+override = false
+```
```

Comment:
As discussed in the WG, a possible alternative to these combinations of `override` & `default` could be to provide a single `priority` parameter. This could either be a generic integer, where the process claiming the highest priority (probably with the lowest number) would become the default process for the app; or a predetermined list of priority strings such as "highest", "high", "medium", "low" & "background".

Not only would this approach provide an overriding highest priority process for the app, but this would also provide a simple way of determining the highest priority process for each process type.

The example above could be rewritten as

```toml
[[processes]]
type = web
command = bundle exec ruby app.rb
priority = 1 // highest

[[processes]]
type = worker
command = bundle exec ruby worker.rb
priority = 4 // low
```